db/versions/0070.yml

```diff
@@ -0,0 +1,180 @@
+        offset get_page_offset(page_offset_in);
+      end
+  update_worker_pool_with_capacity_and_counts_by_state:
+    description: |-
+      Update API-accessible columns on an existig worker pool.  All fields are
+      overridden, but if the provider_id changes, then the existing provider_id
+      is added to previous_provider_ids.  The return value contains values
+      required for an API response and previous_provider_id (singular) containing
+      the provider_id found before the update.  If no such worker pool exists,
+      the return value is an empty set.
+    mode: write
+    serviceName: worker_manager
+    args: worker_pool_id_in text, provider_id_in text, description_in text, config_in jsonb, last_modified_in timestamptz, owner_in text, email_on_error_in boolean
+    returns: table(worker_pool_id text, provider_id text, description text, config jsonb, created timestamptz, last_modified timestamptz, owner text, email_on_error boolean, previous_provider_id text, current_capacity integer, requested_count integer, running_count integer, stopping_count integer, stopped_count integer, requested_capacity integer, running_capacity integer, stopping_capacity integer, stopped_capacity integer)
+    body: |-
+      declare
+        existing record;
+      begin
+        select
+          worker_pools.provider_id,
+          worker_pools.previous_provider_ids
+        from worker_pools
+        where worker_pools.worker_pool_id = worker_pool_id_in
+        -- lock this row for the duration of this transaction..
+        for update
+        into existing;
+
+        -- update previous_provider_ids, if the provider_id has changed
+        if existing.provider_id <> provider_id_in then
+          -- remove both provider IDs to avoid duplicates, then re-add existing.provider_id
+          existing.previous_provider_ids = (existing.previous_provider_ids - provider_id_in - existing.provider_id) || jsonb_build_array(existing.provider_id);
+        end if;
+
+        return query update worker_pools
+        set
+          provider_id = provider_id_in,
+          description = description_in,
+          config = config_in,
+          last_modified = last_modified_in,
+          owner = owner_in,
+          email_on_error = email_on_error_in,
+          previous_provider_ids = existing.previous_provider_ids
+        where worker_pools.worker_pool_id = worker_pool_id_in
+        returning
+          worker_pools.worker_pool_id,
+          worker_pools.provider_id,
+          worker_pools.description,
+          worker_pools.config,
+          worker_pools.created,
+          worker_pools.last_modified,
+          worker_pools.owner,
+          worker_pools.email_on_error,
+          existing.provider_id as previous_provider_id,
+          coalesce((
+            select sum(workers.capacity) from workers where
+              workers.worker_pool_id = worker_pool_id_in and
+              workers.state != 'stopped'),
+            0)::integer,
```

Comment:
Hm, this worries me a little with respect to performance -- is it performing 8 additional index scans of the workers table?

Reply:
Afaik there are some difficulties to add a join and grouping in the update return statement.
We can also solve this on the api level, this function is only being used in `POST /worker-pool/:id` method, so we can drop that sub-query here and trigger previously defined new function to get those stats. 

I'm not sure how Postgres internally will handle transaction if we'd introduce join in that update query. Would it need to lock whole `workers` table?

Reply:
That is a good question -- adding load with more scans is bad, but locking a table is _very_ bad!  I don't think it would lock the table, but it's good to be sure!